### ISSUE TITLE
Custom Columns breaks Pivot Table

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#14604]
**Describe the bug**
Custom Columns from pre-aggregate added to Pivot Table rows section

**To Reproduce**
1. Custom question > Sample Dataset > Orders
2. Custom Column `1+1` as "Should not be included"
3. Summarize Count by People.Source and Products.Category
![image](https://user-images.githubusercontent.com/1447303/106467172-90167a00-649c-11eb-9233-f3c9e39dd4a0.png)
4. Visualize as Pivot Table, noticed the Custom Column is included:
![image](https://user-images.githubusercontent.com/1447303/106467246-aae8ee80-649c-11eb-88a0-e944ae7b03f4.png)

**Information about your Metabase Installation:**
Tested 0.38.0-rc4

**Comment by flamber** [comment]
Changing to P1, since it breaks Pivot Table - and grand totals:
![image](https://user-images.githubusercontent.com/1447303/106469858-1f715c80-64a0-11eb-8765-e843a3ed66bc.png)
![image](https://user-images.githubusercontent.com/1447303/106469922-3617b380-64a0-11eb-9053-907e3d26f974.png)

**Comment by mazameli** [comment]
cc @paulrosenzweig

**Comment by nemanjaglumac** [comment]
@paulrosenzweig did you plan to include a test for this and/or will it help if I make a quick Cypress repro to save you some time?

**Comment by paulrosenzweig** [comment]
The issue is that the pivot endpoints include non-aggregated expressions in the result set.
As a specific example:

```
POST /api/dataset

{"type":"query","query":{"source-table":2,"expressions":{"two":["+",1,["field-id",15]]},"aggregation":[["count"]],"breakout":[["fk->",["field-id",11],["field-id",26]]]},"database":1,"parameters":[]}
```

That request correctly returns two columns. The request below should return one additional column (pivot grouping) but it has two. The “two” column is included when it should be dropped.

```
POST /api/dataset/pivot

{"type":"query","query":{"source-table":2,"expressions":{"two":["+",1,["field-id",15]]},"aggregation":[["count"]],"breakout":[["fk->",["field-id",11],["field-id",26]]]},"database":1,"parameters":[],"pivot_rows":[0],"pivot_cols":[]}
```

It looks like the subtotals missing from the UI are included in the response, but the frontend gets tripped up by the additional column.

**Comment by paulrosenzweig** [comment]
@nemanjaglumac I think a more general Cypress test could be great! The direct issue should be covered by a backend test, but this bug exposed that there's no e2e coverage of custom expressions with pivot tables. A good test of that would aggregate a custom expression and use it in a pivot table. That's slightly different than the repro for this issue which doesn't aggregate the custom expression. However, I think that test will also fail due to this bug, so in some sense it is a repro.

**Comment by nemanjaglumac** [comment]
I ran into one more issue here while playing around with repro steps:

1. Follow the original repro steps
2. For the aggregation add previously custom column (on top of: `Summarize Count by People.Source and Products.Category`) (as Paul suggested)
*Note: at this point there is a console warning: `Unknown field type:  (2) ["expression", "Custom Discount"]`
3. Click visualize (it will show a table)
4. Choose "pivot table" as a visualization - the vizualization will fail at this point
5. The "Play" button overlay makes it impossible to see error details (this has already been documented)
6. Click "Play" icon - nothing changes, but now there is a new console error (or rather errors)
![image](https://user-images.githubusercontent.com/31325167/106515026-59108a80-64d5-11eb-801e-10e5c24447fa.png)
7. Clicking on "Save" question will not actually save the question (although you can click on "Save") and will lead to the same console error from the previous step
8. Click on "Show editor" and remove the Custom column from aggregations
9. The visualization is still broken and it is still impossible to save the question (same console errors still appear)
10. It is also not possible to change visualization type - at this point, UI is completely broken and "locked" - the only thing left to do is to nuke unsaved changes and start again

**Comment by flamber** [comment]
@nemanjaglumac I don't understand step 2. That's what I did in my comment, when I changed priority, and it "works".

**Comment by nemanjaglumac** [comment]
Not sure why it worked for you. Here's the video of my exact repro steps:
https://share.getcloudapp.com/9Zu8P0Xq

**Comment by flamber** [comment]
@nemanjaglumac That's because I was only using 1 metric and 2 dimensions (one was CC), whereas you had 1 metric + 3 dimensions (one was CC)
Definitely worth taking into account. Paul probably has a much better understanding why it's different with just two dimensions.

**Comment by nemanjaglumac** [comment]
Well, it doesn't work with 1 metric and 2 dimensions where one of them is CC. It seems to work initially, but if you **explicitly** select pivot table as a visualization, everything breaks and becomes unusable. Check another video:
https://share.getcloudapp.com/ApuRnROA

**Comment by flamber** [comment]
@nemanjaglumac I have no idea how I got it to work initially - it's exploding just like your example - perhaps I was switching back and forth too much during tests, so it had parts of 0.37.8, which made it work. Dunno, but it's still a P1

**Comment by nemanjaglumac** [comment]
I was using `[Orders.Tax] * 2` as a custom column for this repro, because that is clearly something that is possible to do in pivot tables if we use use UI only:
![image](https://user-images.githubusercontent.com/31325167/106534721-ff1fbd00-64f4-11eb-83b6-90a8191d078f.png)


![ ](https://p58.f1.n0.cdn.getcloudapp.com/items/GGu68KGK/9889fcca-80f2-416e-b1c6-5f4a5bfeef1e.gif?v=f8d6b4fd62832b81f88f368bee1a4653)

**Comment by camsaul** [comment]
Fixed by #14651
